### ISSUE TITLE
Data Scientist can crash a domain node for all users by forcing the RDP Cache to extend infinitely

### ISSUE THREAD

**Issue by IshanMi** [OpenMined/PySyft#6953]

**Comment by IshanMi** [comment]
sigma was set at 0.001 and this created an RDP constant of 41M which would take a LONG time to generate

![Image](https://user-images.githubusercontent.com/32711264/199752273-f16bfef2-4d31-42b8-971a-115d42f90f14.png)

**Comment by shubham3121** [comment]
I had timed it for values in the range 12e6 it was taking like 2.5 hours 😂 for the whole process. I was thinking of solving this in a few ways:
- increase the size of the cache file
- multithreaded the `_get_optimal_alpha_for_constant` computations, as each computation is independent of the other.
- have a background task (via celery) running that increases cache size every few intervals in a separate thread. Although, this may cause memory overload at some time.

**Comment by madhavajay** [comment]
This doesn't seem like something we can quickly fix for 0.7. My suggestion is we revisit this as part of the work to provide the  User access to the Action Graph in 0.8. In the mean time I guess we have to raise an Exception. Forcing a cache rebuild and there by overloading the domain is not really a solution. Also, do we need to extend the entire cache just to solve for a value which is out of range? Can't we just force calculate things outside of the cache but not cache the results?

**Comment by iamtrask** [comment]
There are two quick fixes for this:
1) pre-generate a larger RDP cache (possibly in a thread)
2) the client can deny a .publish() event if it's likely to generate a privacy budget spend higher than the upper bound that the client already knows about.

**Comment by madhavajay** [comment]
I think rather than extending the cache, anything which is outside the cache range should just be passed to `find_alpha_and_eps` sure its going to be slow but its predictably slow where as if you jump really high you generate everything from 1.2 million RDP constants to where ever you jumped inclusive when often you will only want a few values. We can then also introduce a timeout so that if it fails to solve all the RDP Constants for a query by a certain time we raise an Exception. Then at least we arent crashing the server.

Thoughts?

Then for a future task:
- Making that run in parallel in the background will then be easier.
- We can also later change the cache structure so the ranges <50 >50 etc can be seperate arrays. If a user jumps into the distance we can just create a new array at that position, solve the answer and then optionally kick off a multithreaded cache building search around that value range to close the gaps.

**Comment by iamtrask** [comment]
Madhava's solution is the best one. Great thinking sir.

**Comment by iamtrask** [comment]
Note: this is a timing-based side-channel attack we'll need to revisit in the upcoming security push.

**Comment by madhavajay** [comment]
I will split this into multiple tasks on Monday

**Comment by madhavajay** [comment]
The advanced version of this is here: https://github.com/OpenMined/PySyft/issues/7092
